pyro/poutine/trace.py

```diff
@@ -20,26 +20,75 @@ def _warn_if_nan(name, value):
     # Note that -inf log_pdf is fine: it is merely a zero-probability event.
 
 
-class Trace(networkx.DiGraph):
+class DiGraph(networkx.DiGraph):
+    node_dict_factory = collections.OrderedDict
+
+
+class Trace(object):
     """
     Execution trace data structure
     """
 
-    node_dict_factory = collections.OrderedDict
-
     def __init__(self, *args, **kwargs):
         """
         :param string graph_type: string specifying the kind of trace graph to construct
 
         Constructor. Currently identical to networkx.``DiGraph(\*args, \**kwargs)``,
         except for storing the graph_type attribute
         """
+        self._graph = DiGraph(*args, **kwargs)
         graph_type = kwargs.pop("graph_type", "flat")
         assert graph_type in ("flat", "dense"), \
             "{} not a valid graph type".format(graph_type)
         self.graph_type = graph_type
         super(Trace, self).__init__(*args, **kwargs)
 
+    def __del__(self):
```

Comment:
`nx.DiGraph` has a circular reference and would not be gc'd when the `Trace` is collected. This override is needed to cut the circular references in the `nx.DiGraph` object so that it can be immediately gc'd. If you were to remove this line, tests would still pass but `DiGraph` objects would leak. I guess we could strengthen the test to check counts for both `Trace` and `nx.DiGraph` objects, but @karalets already confirmed that things are working correctly.

Reply:
I see. I suppose the gc will detect those cycles at some point, but this should immediately fix the issue. I'm still not sure why the Digraph object should have circular references though, will pick your brains on this later.

Reply:
The cycles arise because `nx.DiGraph` stores view objects and those view objects have references back to the graph, e.g. [here](https://github.com/networkx/networkx/blob/master/networkx/classes/reportviews.py#L335). I've filed https://github.com/networkx/networkx/issues/2885 with a suggested fix.

Reply:
> The cycles arise because nx.DiGraph stores view objects and those view objects have references back to the graph

Kudos to discovering this, and filing the issue! :) Thanks for explaining.